### PR TITLE
refactor(semver): ensure strict null check

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes:

- Fix strict null errors

## Context:

Ref: #12869 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/versioning/semver-coerced/index.ts [hunk 1]
```diff
@@ -11,27 +11,27 @@ export const supportsRanges = false;
 const { is: isStable } = stable;
 
 function sortVersions(a: string, b: string): number {
-  return semver.compare(semver.coerce(a), semver.coerce(b));
+  return semver.compare(semver.coerce(a) ?? a, semver.coerce(b) ?? a);
```
Comment (viceice) [inline lib/versioning/semver-coerced/index.ts:14]: Wrong, return 0 if version can't coerced

File: lib/versioning/semver-coerced/index.ts [hunk 2]
```diff
@@ -11,27 +11,27 @@ export const supportsRanges = false;
...
 }
 
 function getMajor(a: string | SemVer): number | null {
-  return semver.major(semver.coerce(a));
+  return semver.major(semver.coerce(a) ?? a);
```
Comment (viceice) [inline lib/versioning/semver-coerced/index.ts:18]: return  null when version can't coerced

File: lib/versioning/semver-coerced/index.ts [hunk 3]
```diff
@@ -11,27 +11,27 @@ export const supportsRanges = false;
...
 }
 
 function getMinor(a: string | SemVer): number | null {
-  return semver.minor(semver.coerce(a));
+  return semver.minor(semver.coerce(a) ?? a);
```
Comment (viceice) [inline lib/versioning/semver-coerced/index.ts:22]: see above

File: lib/versioning/semver-coerced/index.ts [hunk 4]
```diff
@@ -11,27 +11,27 @@ export const supportsRanges = false;
...
 }
 
 function getPatch(a: string | SemVer): number | null {
   return semver.patch(a);
 }
 
 function matches(version: string, range: string): boolean {
-  return semver.satisfies(semver.coerce(version), range);
+  return semver.satisfies(semver.coerce(version) ?? version, range);
```
Comment (viceice) [inline lib/versioning/semver-coerced/index.ts:30]: return false if coerce fails

File: lib/versioning/semver-coerced/index.ts [hunk 5]
```diff
@@ -11,27 +11,27 @@ export const supportsRanges = false;
...
 }
 
 function equals(a: string, b: string): boolean {
-  return semver.eq(semver.coerce(a), semver.coerce(b));
+  return semver.eq(semver.coerce(a) ?? a, semver.coerce(b) ?? b);
```
Comment (viceice) [inline lib/versioning/semver-coerced/index.ts:34]: return false if coerce fails

File: lib/versioning/semver-coerced/index.ts [hunk 6]
```diff
@@ -44,7 +44,7 @@ function getSatisfyingVersion(
 ): string | null {
   const coercedVersions = versions.map((version) => {
     const coercedVersion = semver.coerce(version);
-    return coercedVersion ? coercedVersion.version : null;
+    return coercedVersion ? coercedVersion.version : version;
```
Comment (viceice) [inline lib/versioning/semver-coerced/index.ts:47]: revert and append  `.filter(Boolean)` instead

File: lib/versioning/semver-coerced/index.ts [hunk 7]
```diff
@@ -55,17 +55,20 @@ function minSatisfyingVersion(
...
   });
   return semver.minSatisfying(coercedVersions, range);
 }
 
 function isLessThanRange(version: string, range: string): boolean {
-  return semver.ltr(semver.coerce(version), range);
+  return semver.ltr(semver.coerce(version) ?? version, range);
```
Comment (viceice) [inline lib/versioning/semver-coerced/index.ts:64]: return false when coerce fails

File: lib/versioning/semver-coerced/index.ts [hunk 8]
```diff
@@ -55,17 +55,20 @@ function minSatisfyingVersion(
...
 }
 
 function isGreaterThan(version: string, other: string): boolean {
-  return semver.gt(semver.coerce(version), semver.coerce(other));
+  return semver.gt(
+    semver.coerce(version) ?? version,
+    semver.coerce(other) ?? version
+  );
```
Comment (viceice) [inline lib/versioning/semver-coerced/index.ts:71]: see above

File: lib/versioning/semver-coerced/index.ts [hunk 9]
```diff
@@ -81,7 +84,8 @@ function isSingleVersion(version: string): string | boolean | null {
 }
 
 // If this is left as an alias, inputs like "17.04.0" throw errors
-export const isVersion = (input: string): string | boolean => isValid(input);
+export const isVersion = (input: string): string | boolean | null =>
```
Comment (viceice) [inline lib/versioning/semver-coerced/index.ts:87]: @rarkins We should change the `VersioningApi` interface to always return bool for `isCompatible`, `isSingleVersion`, `isValid` and `isVersion`

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/versioning/semver-coerced/index.ts (hunk 1 above)
Comment (RahulGautamSingh) [reply lib/versioning/semver-coerced/index.ts:14]: semver.compare only accepts the type `string | semver.SemVer` if we use a 0 there it won't pass the null check.

File: lib/versioning/semver-coerced/index.ts [hunk 10]
```diff
@@ -11,27 +11,35 @@ export const supportsRanges = false;
 const { is: isStable } = stable;
 
 function sortVersions(a: string, b: string): number {
-  return semver.compare(semver.coerce(a), semver.coerce(b));
+  const aCoerced = semver.coerce(a),
+    bCoerced = semver.coerce(b);
```
Comment (viceice) [inline lib/versioning/semver-coerced/index.ts:15]:
```suggestion
  const aCoerced = semver.coerce(a);
  const bCoerced = semver.coerce(b);
```

Clear code style.

File: lib/versioning/semver-coerced/index.ts [hunk 11]
```diff
@@ -11,27 +11,35 @@ export const supportsRanges = false;
...
 function matches(version: string, range: string): boolean {
-  return semver.satisfies(semver.coerce(version), range);
+  const coercedVersion = semver.coerce(version);
+  return coercedVersion ? semver.satisfies(coercedVersion, range) : false;
 }
 
 function equals(a: string, b: string): boolean {
-  return semver.eq(semver.coerce(a), semver.coerce(b));
+  const aCoerced = semver.coerce(a),
+    bCoerced = semver.coerce(b);
```
Comment (viceice) [inline lib/versioning/semver-coerced/index.ts:41]: see other comment

File: lib/versioning/semver-coerced/index.ts [hunk 12]
```diff
@@ -42,30 +50,40 @@ function getSatisfyingVersion(
...
-  const coercedVersions = versions.map((version) => {
-    const coercedVersion = semver.coerce(version);
-    return coercedVersion ? coercedVersion.version : null;
-  });
+  const coercedVersions = versions
+    .map((version) => {
+      const coercedVersion = semver.coerce(version);
+      return coercedVersion ? coercedVersion.version : '';
+    })
+    .filter((version) => version !== '');
```
Comment (viceice) [inline lib/versioning/semver-coerced/index.ts:58]:
```suggestion
  const coercedVersions = versions
    .map((version) => semver.coerce(version))
    .filter(Boolean);
```

File: lib/versioning/semver-coerced/index.ts [hunk 13]
```diff
@@ -42,30 +50,40 @@ function getSatisfyingVersion(
...
```
Comment (viceice) [inline lib/versioning/semver-coerced/index.ts:72]: See other comment

File: lib/versioning/semver-coerced/index.ts [hunk 14]
```diff
@@ -42,30 +50,40 @@ function getSatisfyingVersion(
...
+  return coercedVersion ? semver.ltr(coercedVersion, range) : false;
 }
 
 function isGreaterThan(version: string, other: string): boolean {
-  return semver.gt(semver.coerce(version), semver.coerce(other));
+  const coercedVersion = semver.coerce(version);
+  const coercedOther = semver.coerce(other);
+  return coercedVersion && coercedOther
+    ? semver.gt(coercedVersion, coercedOther)
+    : false;
```
Comment (viceice) [inline lib/versioning/semver-coerced/index.ts:86]: needs coverage

File: lib/versioning/semver-coerced/index.ts (hunk 12 above)
Comment (RahulGautamSingh) [reply lib/versioning/semver-coerced/index.ts]: Applied this:-
```js
const coercedVersions = versions
    .map((version) => {
      const coercedVersion = semver.coerce(version);
      return coercedVersion ? coercedVersion.version : '';
    })
    .filter(Boolean);
    ```
    
    Using` .map((version) => semver.coerce(version))` results in null-check error.

File: lib/versioning/semver-coerced/index.ts [hunk 15]
```diff
@@ -42,30 +50,41 @@ function getSatisfyingVersion(
...
-  const coercedVersions = versions.map((version) => {
-    const coercedVersion = semver.coerce(version);
-    return coercedVersion ? coercedVersion.version : null;
-  });
+  const coercedVersions = versions
+    .map((version) => {
+      const coercedVersion = semver.coerce(version);
+      return coercedVersion ? coercedVersion.version : '';
+    })
+    .filter(Boolean);
```
Comment (viceice) [inline lib/versioning/semver-coerced/index.ts:72]:
```suggestion
    .map((version) => semver.coerce(version))
    .filter<SemVer>(Boolean);
```
should do the typecheck 😉

File: lib/versioning/semver-coerced/index.spec.ts [hunk 16]
```diff
@@ -68,15 +68,19 @@ describe('versioning/semver-coerced/index', () => {
...
+
+    it('should return false if version cannot be coerced', () => {
+      expect(semverCoerced.isGreaterThan('e.e.e', '4.1.0')).toBeFalsy();
+    });
   });
 
   describe('.isLessThanRange(version, range)', () => {
     it('should return true for a lower version in strict semver', () => {
-      expect(semverCoerced.isLessThanRange('1.0.2', '~2.0')).toBeTruthy();
+      expect(semverCoerced.isLessThanRange?.('1.0.2', '~2.0')).toBeTruthy();
```
Comment (rarkins) [inline lib/versioning/semver-coerced/index.spec.ts:79]: Why is this necessary? Isn't it always defined?

Comment (viceice) [reply lib/versioning/semver-coerced/index.spec.ts:79]: It's an optional function
https://github.com/renovatebot/renovate/blob/74d7691befa635c1220d8569936273bfb950e748/lib/versioning/types.ts#L26

File: lib/versioning/semver-coerced/index.ts (hunk 9 above)
Comment (rarkins) [reply lib/versioning/semver-coerced/index.ts:103]: Do we need to do so before continuing with these versioning checks?

Comment (viceice) [reply lib/versioning/semver-coerced/index.ts:103]: i think it's independend

File: lib/versioning/semver-coerced/index.ts (hunk 15 above)
Comment (RahulGautamSingh) [reply lib/versioning/semver-coerced/index.ts:72]: It won't work cz the `coercedVersions` array needs to be of type `string[]` and not `SemVer[]`.

Comment (RahulGautamSingh) [reply lib/versioning/semver-coerced/index.ts:72]: Can use this instead:-
```js
  type ExcludesFalse = <T>(x: T | false | undefined | null) => x is T;
  const coercedVersions = versions
    .map((version) => semver.coerce(version)?.version)
    .filter(Boolean as any as ExcludesFalse);
```

Comment (viceice) [reply lib/versioning/semver-coerced/index.ts:72]: ok, found a better way, see updated suggestion

File: lib/versioning/semver-coerced/index.ts (hunk 9 above)
Comment (RahulGautamSingh) [reply lib/versioning/semver-coerced/index.ts:103]: https://github.com/renovatebot/renovate/blob/1445fb07bf1f530bb635e66b3521a1966ec9712f/lib/versioning/ubuntu/index.ts#L14-L19

Is the type-check necessary when the argument type is already declared?

File: lib/versioning/semver-coerced/index.ts [hunk 17]
```diff
@@ -42,30 +51,35 @@ function getSatisfyingVersion(
   versions: string[],
   range: string
 ): string | null {
-  const coercedVersions = versions.map((version) => {
-    const coercedVersion = semver.coerce(version);
-    return coercedVersion ? coercedVersion.version : null;
-  });
+  const coercedVersions: string[] = versions
```
Comment (viceice) [inline lib/versioning/semver-coerced/index.ts:54]: Is the explicit type really needed?

Comment (RahulGautamSingh) [reply lib/versioning/semver-coerced/index.ts:54]: Nope, not necessary.

Comment (viceice) [reply lib/versioning/semver-coerced/index.ts:54]:
```suggestion
  const coercedVersions = versions
```

File: lib/versioning/semver-coerced/index.ts [hunk 18]
```diff
@@ -42,30 +51,35 @@ function getSatisfyingVersion(
...
 
 function minSatisfyingVersion(
   versions: string[],
   range: string
 ): string | null {
-  const coercedVersions = versions.map((version) => {
-    const coercedVersion = semver.coerce(version);
-    return coercedVersion ? coercedVersion.version : null;
-  });
+  const coercedVersions: string[] = versions
```
Comment (viceice) [inline lib/versioning/semver-coerced/index.ts:65]:
```suggestion
  const coercedVersions = versions
```

**Review by viceice — APPROVED** [review]
